src/python/pants/rules/core/run.py

```diff
@@ -1,35 +1,58 @@
 # Copyright 2019 Pants project contributors (see CONTRIBUTORS.md).
 # Licensed under the Apache License, Version 2.0 (see LICENSE).
 
+from pathlib import Path
+
+from pants.build_graph.address import Address
 from pants.engine.addressable import BuildFileAddresses
 from pants.engine.console import Console
+from pants.engine.fs import DirectoryToMaterialize, Workspace
 from pants.engine.goal import Goal
 from pants.engine.interactive_runner import InteractiveProcessRequest, InteractiveRunner
 from pants.engine.rules import console_rule
+from pants.engine.selectors import Get
+from pants.rules.core.binary import CreatedBinary
+from pants.util.contextutil import temporary_dir
 
 
 class Run(Goal):
   """Runs a runnable target."""
-  name = 'v2-run'
+  name = 'run'
 
 
+#TODO(gregs) - the `run` rule should really only accept a single target, but we don't have the infrastructure
+# yet to support a console rule that can request one and only one target (rather than BuildFileAddresses plural
+# or Specs plural). 
 @console_rule
-def run(console: Console, runner: InteractiveRunner, build_file_addresses: BuildFileAddresses) -> Run:
-  console.write_stdout("Running the `run` goal\n")
-
-  request = InteractiveProcessRequest(
-    argv=["/usr/bin/python"],
-    env=("TEST_ENV", "TEST"),
-    run_in_workspace=False,
-  )
-
-  try:
-    res = runner.run_local_interactive_process(request)
-    print(f"Subprocess exited with result: {res.process_exit_code}")
-    yield Run(res.process_exit_code)
-  except Exception as e:
-    print(f"Exception when running local interactive process: {e}")
-    yield Run(-1)
+def run(console: Console, workspace: Workspace, runner: InteractiveRunner, addresses: BuildFileAddresses) -> Run:
+  bfa = addresses.dependencies[0]
+  target = bfa.to_address()
+  binary = yield Get(CreatedBinary, Address, target)
+
+  with temporary_dir(cleanup=True) as tmpdir:
+    dirs_to_materialize = (DirectoryToMaterialize(path=str(tmpdir), directory_digest=binary.digest),)
+    workspace.materialize_directories(dirs_to_materialize)
```

Comment:
This is pretty odd. The point of the `workspace` type was to allow for materializing inside of the "workspace"/buildroot: allowing arbitrary paths to be passed here seems overly powerful and more complicated than necessary.

Should an `InteractiveProcessRequest` take the digest and materialize it somewhere on its own?

Reply:
If we wanted to constrain `Workspace` to only being capable of writing to the buildroot, then yeah we'd have to allow `InteractiveProcessRequest` to materialize its own `Digest`. I'll add an issue for this.